pkg/infra/tower.go

```diff
@@ -0,0 +1,121 @@
+package infra
+
+import (
+	"log"
+	"math/rand"
+
+	"github.com/SOMAS2021/SOMAS2021/pkg/messages"
+)
+
+type Tower struct {
+	currPlatFood    float64
+	maxPlatFood     float64
+	currPlatFloor   uint64
+	agentCount      int
+	agents          map[string]*Base
+	agentsPerFloor  int
+	ticksPerDay     int
+	missingAgents   map[int][]int // key: floor, value: types of missing agents
+	reshufflePeriod int
+	tickCounter     int
+}
+
+func NewTower(currPlatFood float64, currPlatFloor uint64, agentCount,
+	agentsPerFloor, ticksPerDay, reshufflePeriod int) *Tower {
+	t := &Tower{
+		currPlatFood:    currPlatFood,
+		maxPlatFood:     currPlatFood,
+		currPlatFloor:   currPlatFloor,
+		agentCount:      agentCount,
+		agents:          make(map[string]*Base),
+		agentsPerFloor:  agentsPerFloor,
+		ticksPerDay:     ticksPerDay,
+		missingAgents:   make(map[int][]int),
+		reshufflePeriod: reshufflePeriod,
+		tickCounter:     1,
+	}
+	return t
+}
+
+func (t *Tower) Tick() {
+	//logs
+	log.Printf("A log from the tower! Tick no: %d", t.tickCounter)
+	log.Printf("The food left on the platform = %f", t.currPlatFood)
+
+	//useful parameters
+	day := 24 * 60
+	numOfFloors := t.agentCount / int(t.agentsPerFloor)
+	platformMovePeriod := day / numOfFloors // can add min/max
+
+	if (t.tickCounter)%(t.reshufflePeriod) == 0 {
+		t.reshuffle(numOfFloors)
+	}
+	if (t.tickCounter)%(platformMovePeriod) == 0 {
+		t.currPlatFloor++
+	}
+	if (t.tickCounter)%(day) == 0 {
+		t.hpDecay() // deacreases HP and kills if < 0
+		t.ResetTower()
+	}
+
+	t.tickCounter++
+}
+
+func (t *Tower) GetMissingAgents() map[int][]int {
+	deadAgents := t.missingAgents
+	t.missingAgents = make(map[int][]int)
+	return deadAgents
+}
+
+func (t *Tower) AddAgent(bagent *Base) {
+	t.agents[bagent.id] = bagent
+}
+
+func (t *Tower) reshuffle(numOfFloors int) {
+	remainingVacancies := make([]int, numOfFloors)
+	log.Printf("Reshuffling alive agents...")
+	log.Printf("Number of agents: %d", len(t.agents))
+	for i := 0; i < numOfFloors; i++ { // adding a max to each floor
+		remainingVacancies[i] = t.agentsPerFloor
+	}
+	// allocating agents to floors randomly
+	// iterate through the uuid strings of each agent
+	for _, agent := range t.agents {
+		newFloor := rand.Intn(numOfFloors)
+		for remainingVacancies[newFloor] == 0 {
+			newFloor = rand.Intn(numOfFloors)
+		}
+		agent.setFloor(newFloor + 1)
+		remainingVacancies[newFloor]--
+	}
+}
+
+func (t *Tower) hpDecay() {
+	// TODO: can add a parameter
+	for _, agent := range t.agents {
+		newHP := agent.HP() - 20
```

Comment:
Parametrise this decay

Reply:
Yeah that's why there is the TODO 😄 Feel free to make an issue for it! Agent teams can PR on top of this function

Reply:
Apologies, skipped over that comment.